File: widget/YearProgress.cs
using System;

namespace EverlastimerWidget;

/// <summary>
/// Pure date-math for year progress. Deliberately mirrors the logic in
/// the Flutter app's lib/core/year_progress.dart so the two stay in sync
/// conceptually, even though they're two separate codebases with no
/// runtime connection to each other.
/// </summary>
public sealed class YearProgress
{
    public DateTime Now { get; }
    public int Year { get; }
    public DateTime StartOfYear { get; }
    public DateTime StartOfNextYear { get; }

    private YearProgress(DateTime now)
    {
        Now = now;
        Year = now.Year;
        StartOfYear = new DateTime(Year, 1, 1);
        StartOfNextYear = new DateTime(Year + 1, 1, 1);
    }

    public static YearProgress FromNow() => new(DateTime.Now);

    public bool IsLeapYear => DateTime.IsLeapYear(Year);

    public int TotalDaysInYear => IsLeapYear ? 366 : 365;

    private TimeSpan Elapsed => Now - StartOfYear;
    private TimeSpan TotalYearDuration => StartOfNextYear - StartOfYear;

    /// 0.0–1.0 fraction of the year elapsed, sub-day precision so the ring
    /// advances smoothly rather than jumping once a day.
    public double Fraction
    {
        get
        {
            var f = Elapsed.TotalSeconds / TotalYearDuration.TotalSeconds;
            return Math.Clamp(f, 0.0, 1.0);
        }
    }

    public double PercentComplete => Fraction * 100.0;

    public int DaysCompleted => (int)Elapsed.TotalDays;
    public int DaysRemaining => TotalDaysInYear - DaysCompleted;

    public int CurrentMonth => Now.Month;

    public int CurrentWeek
    {
        get
        {
            var dayOfYear = DaysCompleted + 1;
            return ((dayOfYear - 1) / 7) + 1;
        }
    }

    public int TotalWeeksInYear => (int)Math.Ceiling(TotalDaysInYear / 7.0);

    public int CurrentDayOfYear => DaysCompleted + 1;

    public long HoursCompleted => (long)Elapsed.TotalHours;
    public int TotalHoursInYear => TotalDaysInYear * 24;
}

File: app/lib/core/year_progress.dart
/// Pure date-math for computing how far through the current year we are.
///
/// Deliberately has zero Flutter/UI imports — this is the kind of logic
/// that's easy to unit test and easy to reuse from a home-screen widget,
/// a native widget bridge, or a background isolate.
library;

class YearProgress {
  final DateTime now;
  final int year;
  final DateTime startOfYear;
  final DateTime startOfNextYear;

  YearProgress._({
    required this.now,
    required this.year,
    required this.startOfYear,
    required this.startOfNextYear,
  });

  /// Builds a [YearProgress] snapshot. Pass [now] explicitly in tests;
  /// defaults to [DateTime.now] in production.
  factory YearProgress.fromDate(DateTime? now) {
    final effectiveNow = now ?? DateTime.now();
    final year = effectiveNow.year;
    return YearProgress._(
      now: effectiveNow,
      year: year,
      startOfYear: DateTime(year, 1, 1),
      startOfNextYear: DateTime(year + 1, 1, 1),
    );
  }

  bool get isLeapYear =>
      (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);

  int get totalDaysInYear => isLeapYear ? 366 : 365;

  Duration get _elapsed => now.difference(startOfYear);
  Duration get _totalYearDuration => startOfNextYear.difference(startOfYear);

  /// 0.0–1.0 fraction of the year elapsed, using sub-day precision
  /// (so the ring advances smoothly, not just once a day).
  double get fraction =>
      (_elapsed.inSeconds / _totalYearDuration.inSeconds).clamp(0.0, 1.0);

  /// Percentage 0.0–100.0, one decimal place expected at display time.
  double get percentComplete => fraction * 100;

  int get daysCompleted => _elapsed.inDays;
  int get daysRemaining => totalDaysInYear - daysCompleted;

  int get currentMonth => now.month; // 1–12

  int get currentWeek {
    // ISO-8601-ish: week number based on days elapsed since Jan 1.
    final dayOfYear = daysCompleted + 1;
    return ((dayOfYear - 1) / 7).floor() + 1;
  }

  int get totalWeeksInYear => (totalDaysInYear / 7).ceil();

  int get currentDayOfYear => daysCompleted + 1;

  int get hoursCompleted => _elapsed.inHours;
  int get totalHoursInYear => totalDaysInYear * 24;
}